lib/TPP/TileConsumerAndFuseProducers.cpp

```diff
@@ -25,108 +27,92 @@ using namespace mlir;
-  if (!offsetAsInt)
-    return false;
-  Optional<int64_t> sizeAsInt = getConstantIntValue(loopRange.size);
-  if (!sizeAsInt)
-    return false;
-  Optional<int64_t> strideAsInt = getConstantIntValue(loopRange.stride);
-  return static_cast<bool>(strideAsInt);
+// TODO: improve checks here.
+static bool isBlockedMatmul(linalg::GenericOp genericOp) {
+  return tpp::utils::hasMatmulBody(genericOp);
 }
 
-static int64_t getSizeRange(Range loopRange) {
-  assert(isStaticRange(loopRange));
-  return *getConstantIntValue(loopRange.size) -
-         *getConstantIntValue(loopRange.offset);
+static bool isMatmulLike(Operation *op) {
+  if (isa_and_nonnull<linalg::MatmulOp>(op))
+    return true;
+  if (linalg::GenericOp maybeMatmul = dyn_cast_or_null<linalg::GenericOp>(op))
+    return isBlockedMatmul(maybeMatmul);
+  return false;
 }
 
-static LogicalResult tileDivideIterationDomain(linalg::GenericOp linalgOp,
-                                               ArrayRef<int64_t> tiles,
-                                               OpBuilder &builder) {
-  if (linalgOp.getNumLoops() != tiles.size())
-    return failure();
-  SmallVector<Range> iterationDomain =
-      cast<TilingInterface>(linalgOp.getOperation())
-          .getIterationDomain(builder);
-  for (const auto &it : llvm::enumerate(iterationDomain)) {
-    // fine, we are not tiling along this dimension.
-    if (tiles[it.index()] == 0)
-      continue;
-    // require static loop range
-    if (!isStaticRange(it.value()))
-      return failure();
-    int64_t sizeRange = getSizeRange(it.value());
-    // fail if the tail size equals the range
-    // or is not a full tile.
-    if (tiles[it.index()] == sizeRange)
-      return failure();
-    if (sizeRange % tiles[it.index()] != 0)
-      return failure();
+static bool hasMatmulLikeProducer(linalg::LinalgOp linalgOp) {
+  for (Value operand : linalgOp->getOperands()) {
+    Operation *op = operand.getDefiningOp<linalg::LinalgOp>();
+    if (op && isMatmulLike(op))
+      return true;
   }
-  return success();
+  return false;
 }
 
-struct FuseGenericOp : public OpRewritePattern<linalg::GenericOp> {
-  FuseGenericOp(MLIRContext *context, ArrayRef<int64_t> tileSizes,
-                PatternBenefit benefit = 1)
-      : OpRewritePattern<linalg::GenericOp>(context, benefit),
-        tileSizes(tileSizes) {}
+static FailureOr<scf::SCFTilingResult>
+tileConsumer(RewriterBase &rewriter, TilingInterface consumer,
+             ArrayRef<int64_t> tileSizes) {
+  // If tileSizes is provided use them otherwise defaul to 1, 1, meaning fuse
+  // entire the two outermost loops.
+  SmallVector<int64_t> actualTileSizes = tileSizes.empty()
+                                             ? SmallVector<int64_t>{1, 1}
+                                             : llvm::to_vector(tileSizes);
+  auto options = scf::SCFTilingOptions().setTileSizes(actualTileSizes);
+  FailureOr<scf::SCFTilingResult> tilingResult =
```

Comment:
Don't we need to check if the number of tiles sizes is compatible with the tiling we're trying to do? Say, if you pass more sizes than there are dimensions?

Reply:
If the tile sizes are not correct we should not get there. I will put an assert.